app/src/main/java/com/concordium/wallet/ui/walletconnect/WalletConnectViewModel.kt

```diff
@@ -1364,31 +1404,79 @@ private constructor(
         return sessionRequestIdentityProofIdentities[account.identityId]
     }
 
-    private suspend fun signRequestedMessage(accountKeys: AccountData) {
-        val signMessageInput = SignMessageInput(
-            address = sessionRequestAccount.address,
-            message = sessionRequestMessageToSign,
-            keys = accountKeys,
-        )
-        val signMessageOutput = App.appCore.cryptoLibrary.signMessage(signMessageInput)
-        if (signMessageOutput == null) {
-            Log.e("failed_signing_message")
+    private fun signRequestedMessage(accountKeys: AccountData) {
+        val signatureHex = try {
+            val signKeyHex =
+                App.appCore.gson.fromJson(accountKeys.keys.json, AccountDataKeys::class.java)
+                    .level0
+                    .keys
+                    .keys
+                    .signKey
+
+            val message = when (val signMessageParams = sessionRequestSignMessageParams) {
+                is SignMessageParams.Binary ->
+                    signMessageParams.data.hexToBytes()
+
+                is SignMessageParams.Text ->
+                    signMessageParams.data.encodeToByteArray()
+            }
+
+            // See https://github.com/Concordium/concordium-base/blob/main/mobile_wallet/src/lib.rs#L89
+            val input = SHA256.hash(
+                AccountAddress.from(sessionRequestAccount.address).bytes
+                        + ByteArray(8)
+                        + message
+            )
+
+            ED25519SecretKey.from(signKeyHex)
+                .sign(input)
+                .toHex()
```

Comment:
It seems like you are only signing using one of the keys of the account, when an account might have multiple credentials with multiple keys.
The wallet probably doesn't support multi signature accounts, but when we get there, this will become a bug so might as well deal with it already

Reply:
Indeed, but it seems "single account – single key" principle is deeply embedded into the wallet structures and methods, I don't even know how to iterate account keys. I'm not going to do anything on this, at least in CryptoX. When adding message signatures to the SDK, I'll consider having multiple keys.